staging/src/k8s.io/code-generator/cmd/import-boss/main.go

```diff
@@ -21,32 +21,565 @@ import (
+	for _, imp := range uniq(direct, indirect) {
+		if unmassage(imp) == realPkgPath {
+			// Tests in package "foo_test" depend on the test package for
+			// "foo" (if both exist in a giver directory).
+			continue
+		}
+		klog.V(4).Infof("considering import %q %s %q", pkg.PkgPath, relate(imp), imp)
+		matched := false
+		decided := false
+		for i, file := range restrictionFiles {
+			klog.V(4).Infof("rules file %s", file.path)
+			for j, rule := range file.Rules {
+				if !rule.Transitive && !isDirect[imp] {
+					continue
+				}
+				matching := selectors[i][j].MatchString(imp)
+				if !matching {
+					continue
+				}
+				matched = true
+				klog.V(6).Infof("selector %v matches %q", rule.SelectorRegexp, imp)
+
+				disp := rule.Evaluate(imp)
+				if disp == DepAllowed {
+					decided = true
+					break // no further rules, next file
+				} else if disp == DepForbidden {
+					errs = append(errs, fmt.Errorf("%q %s %q is forbidden by %s", pkg.PkgPath, relate(imp), imp, file.path))
+					decided = true
+					break // no further rules, next file
+				}
+			}
+			if decided {
+				break // no further files, next import
+			}
+		}
+		if matched && !decided {
+			klog.V(5).Infof("%q %s %q did not match any rule", pkg, relate(imp), imp)
+			errs = append(errs, fmt.Errorf("%q %s %q did not match any rule", pkg.PkgPath, relate(imp), imp))
+		}
+	}
+
+	if len(errs) > 0 {
+		return errs
+	}
+
+	return nil
+}
+
+func uniq(slices ...[]string) []string {
+	m := map[string]bool{}
+	for _, sl := range slices {
+		for _, str := range sl {
+			m[str] = true
+		}
+	}
+	ret := []string{}
+	for str := range m {
+		ret = append(ret, str)
+	}
+	sort.Strings(ret)
+	return ret
+}
+
+func hasPathPrefix(path, prefix string) bool {
+	if prefix == "" || path == prefix {
+		return true
+	}
+	if !strings.HasSuffix(path, string(filepath.Separator)) {
+		prefix = prefix + string(filepath.Separator)
+	}
+	return strings.HasPrefix(path, prefix)
+}
+
+func transitiveImports(pkg *packages.Package) ([]string, []string) {
+	direct := []string{}
+	indirect := []string{}
+	seen := map[string]bool{}
+	for _, imp := range pkg.Imports {
+		direct = append(direct, imp.PkgPath)
+		dfsImports(&indirect, seen, imp)
+	}
+	return direct, indirect
+}
+
+func dfsImports(dest *[]string, seen map[string]bool, p *packages.Package) {
+	for _, p2 := range p.Imports {
+		if seen[p2.PkgPath] {
+			continue
+		}
+		seen[p2.PkgPath] = true
+		*dest = append(*dest, p2.PkgPath)
+		dfsImports(dest, seen, p2)
+	}
+}
+
+// verifyInverseRules checks that all packages that import a package are allowed to import it.
+func (boss *ImportBoss) verifyInverseRules(pkg *packages.Package, restrictionFiles []*FileFormat) []error {
+	klog.V(3).Infof("verifying pkg %q inverse-rules", pkg.PkgPath)
+
+	// compile all Selector regex in all restriction files
+	selectors := make([][]*regexp.Regexp, len(restrictionFiles))
+	for i, restrictionFile := range restrictionFiles {
+		for _, r := range restrictionFile.InverseRules {
```

Comment:
I couldn't help but notice how closely this function resembles `verifyRules`.  Is it worth trying to consolidate down to a common function?  I was able to eyeball some of the differences, so I can see arguments both ways, but figured I should ask.

Reply:
(Oh, I forgot to check if this is pre-existing code that got moved, if so please ignore...)

Reply:
I did make sure they are as close to the same as possible, but I did not try to factor it away.